File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/opengl/model/GLRect.java
package br.com.insanitech.spritekit.opengl.model;

import br.com.insanitech.spritekit.core.ValueAssign;

/**
 * Created by anderson on 7/3/15.
 */
public class GLRect implements ValueAssign<GLRect> {
    public GLPoint origin;
    public GLSize size;

    public static GLRect centerRect(GLPoint point, GLSize size) {
        return new GLRect(point.x - size.width / 2, point.y - size.height / 2, size.width, size.height);
    }

    public GLRect() {
        this(0, 0, 0, 0);
    }

    public GLRect(float x, float y, float width, float height) {
        origin = new GLPoint(x, y);
        size = new GLSize(width, height);
    }

    public GLRect(GLPoint point, GLSize size) {
        this.origin = point;
        this.size = size;
    }

    public float getX() {
        return origin.x;
    }

    public float getY() {
        return origin.y;
    }

    public float getWidth() {
        return size.width;
    }

    public float getHeight() {
        return size.height;
    }

    public boolean containsPoint(GLPoint point) {
        return point.x > origin.x &&
                point.y > origin.y &&
                point.x < origin.x + size.width &&
                point.y < origin.y + size.height;
    }

    @Override
    public void assignByValue(GLRect other) {
        origin.assignByValue(other.origin);
        size.assignByValue(other.size);
    }
}

File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/opengl/model/GLRect.kt
package br.com.insanitech.spritekit.opengl.model

import br.com.insanitech.spritekit.core.ValueAssign

/**
 * Created by anderson on 7/3/15.
 */
internal class GLRect() : ValueAssign<GLRect> {
    var origin: GLPoint = GLPoint()
        set(value) { this.origin.assignByValue(value) }
    var size: GLSize = GLSize()
        set(value) { this.size.assignByValue(value) }

    constructor(x: Float, y: Float, width: Float, height: Float) : this() {
        this.origin.x = x
        this.origin.y = y
        this.size.width = width
        this.size.height = height
    }

    constructor(point: GLPoint, size: GLSize) : this() {
        this.origin = point
        this.size = size
    }

    val x: Float
        get() = this.origin.x

    val y: Float
        get() = this.origin.y

    val width: Float
        get() = this.size.width

    val height: Float
        get() = this.size.height

    fun containsPoint(point: GLPoint): Boolean {
        return point.x > this.origin.x &&
                point.y > this.origin.y &&
                point.x < this.origin.x + this.size.width &&
                point.y < this.origin.y + this.size.height
    }

    override fun assignByValue(other: GLRect) {
        this.origin.assignByValue(other.origin)
        this.size.assignByValue(other.size)
    }

    companion object {
        fun centerRect(point: GLPoint, size: GLSize): GLRect =
                GLRect(point.x - size.width / 2, point.y - size.height / 2, size.width, size.height)
    }
}
